conans/test/functional/graph_lock/graph_lock_test.py

```diff
@@ -387,6 +387,77 @@ def test_base_options(self):
         pkg_lock = client.load("pkg.lock")
         self.assertIn('"options": "shared=False"', pkg_lock)
 
+    def test_config_option(self):
+        # https://github.com/conan-io/conan/issues/7991
+        client = TestClient()
+        pahomqttc = textwrap.dedent("""
+            from conans import ConanFile
+            class PahoMQTCC(ConanFile):
+                options = {"shared": [True, False]}
+                default_options = {"shared": False}
+
+                def config_options(self):
+                    # This is weaker than "configure()", will be overwritten by downstream
+                    self.options.shared = True
+            """)
+
+        pahomqttcpp = textwrap.dedent("""
+            from conans import ConanFile
+            class Meta(ConanFile):
+                requires = "pahomqttc/1.0"
+                def configure(self):
+                    self.options["pahomqttc"].shared = False
+            """)
+
+        client.save({"pahomqttc/conanfile.py": pahomqttc,
+                     "pahomqttcpp/conanfile.py": pahomqttcpp,
+                     "consumer/conanfile.txt": "[requires]\npahomqttcpp/1.0"})
+        client.run("export pahomqttc pahomqttc/1.0@")
+        client.run("export pahomqttcpp pahomqttcpp/1.0@")
+
+        client.save({"conanfile.txt": "[requires]\npahomqttcpp/1.0"}, clean_first=True)
+        client.run("install . --build -o paho-mqtt-c:shared=False")
```

Comment:
You are creating `conanfile.txt` twice

```suggestion
        client.run("install consumer/conanfile.txt --build -o paho-mqtt-c:shared=False")
```